srml/staking/src/slashing.rs

```diff
@@ -0,0 +1,772 @@
+		} else {
+			Zero::zero()
+		};
+
+		if !reward.is_zero() {
+			changed = true;
+			span_record.paid_out += reward;
+			*self.paid_out += reward;
+		}
+
+		if changed {
+			self.dirty = true;
+			<Module<T> as Store>::SpanSlash::insert(&span_slash_key, &span_record);
+		}
+
+		Some(target_span.index)
+	}
+}
+
+impl<'a, T: 'a + Trait> Drop for InspectingSpans<'a, T> {
+	fn drop(&mut self) {
+		// only update on disk if we slashed this account.
+		if !self.dirty { return }
+
+		if let Some((start, end)) = self.spans.prune(self.window_start) {
+			for span_index in start..end {
+				<Module<T> as Store>::SpanSlash::remove(&(self.stash.clone(), span_index));
+			}
+		}
+
+		<Module<T> as Store>::SlashingSpans::insert(self.stash, &self.spans);
+
+		// do this AFTER updating span information because it is possible that the
+		// slash may zero the account and cause garbage collection. any stored data
+		// after that point would be orphaned.
+		if let Some(slash) = self.deferred_slash.take() {
+			apply_slash::<T>(self.stash, slash, self.paid_out, self.slash_imbalance)
+		}
+	}
+}
+
+/// Clear slashing metadata for an obsolete era.
+pub(crate) fn clear_era_metadata<T: Trait>(obsolete_era: EraIndex) {
+	<Module<T> as Store>::ValidatorSlashInEra::remove_prefix(&obsolete_era);
+	<Module<T> as Store>::NominatorSlashInEra::remove_prefix(&obsolete_era);
+}
+
+/// Clear slashing metadata for a dead account.
+pub(crate) fn clear_stash_metadata<T: Trait>(stash: &T::AccountId) {
+	let spans = match <Module<T> as Store>::SlashingSpans::take(stash) {
+		None => return,
+		Some(s) => s,
+	};
+
+	// kill slashing-span metadata for account.
+	//
+	// this can only happen while the account is staked _if_ they are completely slashed.
+	// in that case, they may re-bond, but it would count again as span 0. Further ancient
+	// slashes would slash into this new bond, since metadata has now been cleared.
+	for span in spans.iter() {
+		<Module<T> as Store>::SpanSlash::remove(&(stash.clone(), span.index));
+	}
+}
+
+// apply the slash to a stash account, deducting any missing funds from the reward
+// payout, saturating at 0. this is mildly unfair but also an edge-case that
+// can only occur when overlapping locked funds have been slashed.
+fn apply_slash<T: Trait>(
+	stash: &T::AccountId,
+	value: BalanceOf<T>,
+	reward_payout: &mut BalanceOf<T>,
+	slash_imbalance: &mut NegativeImbalanceOf<T>,
+) {
+	let controller = match <Module<T>>::bonded(stash) {
+		None => return, // defensive: should always exist.
+		Some(c) => c,
+	};
+
+	let mut ledger = match <Module<T>>::ledger(&controller) {
+		Some(ledger) => ledger,
+		None => return, // nothing to do.
+	};
+
+	let mut value = value.min(ledger.active);
+
+	if !value.is_zero() {
+		ledger.active -= value;
+
+		// Avoid there being a dust balance left in the staking system.
+		if ledger.active < T::Currency::minimum_balance() {
+			value += ledger.active;
+			ledger.active = Zero::zero();
+		}
+
+		ledger.total -= value;
```

Comment:
We are sure it won't overflow, right? `ledger.total >= ledger.active`?

Reply:
yeah, that's an invariant of the data structure.